File: backend/internal/domain/action_origin.go
package domain

import "context"

type ActionOrigin string

const (
	ActionOriginUser     ActionOrigin = "user"
	ActionOriginMCPAgent ActionOrigin = "mcp_agent"
)

func (origin ActionOrigin) OrUser() ActionOrigin {
	if origin == "" {
		return ActionOriginUser
	}

	return origin
}

type actionOriginContextKey string

const ActionOriginContextKey actionOriginContextKey = "action_origin"

func WithActionOrigin(ctx context.Context, origin ActionOrigin) context.Context {
	return context.WithValue(ctx, ActionOriginContextKey, origin.OrUser())
}

func ActionOriginFromContext(ctx context.Context) ActionOrigin {
	origin, ok := ctx.Value(ActionOriginContextKey).(ActionOrigin)
	if !ok {
		return ActionOriginUser
	}

	return origin.OrUser()
}

File: frontend/src/features/tasks/types/action-origin.ts
export type ActionOrigin = 'user' | 'mcp_agent';

export const normalizeActionOrigin = (origin?: ActionOrigin | null): ActionOrigin => {
  if (origin === 'mcp_agent') {
    return 'mcp_agent';
  }

  return 'user';
};

export const isMCPAgentAction = (origin?: ActionOrigin | null): boolean => {
  return normalizeActionOrigin(origin) === 'mcp_agent';
};
